clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -106,13 +107,21 @@ public void configure(TransportLayer transportLayer, PrincipalBuilder principalB
 
             setSaslState(handshakeRequestEnable ? SaslState.SEND_HANDSHAKE_REQUEST : SaslState.INITIAL);
 
-            // determine client principal from subject.
-            if (!subject.getPrincipals().isEmpty()) {
-                Principal clientPrincipal = subject.getPrincipals().iterator().next();
-                this.clientPrincipalName = clientPrincipal.getName();
-            } else {
-                clientPrincipalName = null;
+            // determine client principal from subject for Kerberos. For other mechanisms, the authenticated
+            // principal (username for PLAIN and SCRAM) is used as authorization id.
+            if (mechanism.equals(SaslConfigs.GSSAPI_MECHANISM)) {
+                try {
+                    Principal clientPrincipal = subject.getPrincipals().iterator().next();
```

Comment:
Thanks for the explanation.

The iterator returned by synchronized connections has to be synchronized by the callers. The code of `SynchronizedCollection` that seems to be used by `getPrincipals`:

```java
public Iterator<E> iterator() {
    return c.iterator(); // Must be manually synched by user!
}
```

So, it looks like we don't really have a choice here.